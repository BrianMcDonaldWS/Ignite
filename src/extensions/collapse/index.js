import container from 'markdown-it-container';

const regExp = /collapse/;

const collapse = md => ({
  validate(params) {
    return params.trim().match(regExp);
  },

  render(tokens, idx) {
    const m = tokens[idx].info.trim().match(/^collapse\s+(.*)$/);

    if (tokens[idx].nesting === 1) {
      const isOpen = m[1].split(' ')[0] === 'open';
      let [, rest] = m;

      if (isOpen) {
        rest = m[1]
          .split(' ')
          .slice(1)
          .join(' ');
      }

      return `
        <details ${isOpen ? 'open' : ''}>
          <summary style=!{!{display: 'flex',alignItems: 'center', padding: 7, position: 'relative'!}!} className="menu-label">
            ${md.renderInline(rest)}
          </summary>
      `;
    }

    return `
      </details>
    `;
  }
});

export default function collapseBlock(md) {
  md.use(container, 'collapse', collapse(md));
}
