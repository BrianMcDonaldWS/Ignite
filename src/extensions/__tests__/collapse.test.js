import MarkdownIt from 'markdown-it';
import collapse from '../collapse';

const md = new MarkdownIt().use(collapse);

test('collapse', () => {
  expect(
    md.render(`
::: collapse

# Title

## I can contain various combinations of elements.
:::
  `)
  ).toMatchSnapshot();

  expect(
    md.render(`
::: collapse open

# Title

## I can contain various combinations of elements.
:::
  `)
  ).toMatchSnapshot();

  expect(
    md.render(`
* ::: collapse [:package: Publishing](./pages/Publishing.html)
  * [Setup](./pages/Publishing.html#setup-branch)
  * [Continuous Integration](./pages/Publishing.html#Continuous-integration)
  `)
  ).toMatchSnapshot();
});
