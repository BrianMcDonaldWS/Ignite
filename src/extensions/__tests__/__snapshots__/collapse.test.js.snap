// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collapse 1`] = `
"
        <details
          open=!{
            undefined ||
            [].find(link => (
              link === ((this && this.props) || props).currentPage ||
              (((this && this.props) || props).currentPage === process.env.baseURL &&
                link === window.configuration.currentFirstPage.replace('.md', '.html'))
              )
            )
          !}
        >
          <summary className='collapse-summary'>
            <p></p>
          </summary>
      <h1>Title</h1>
<h2>I can contain various combinations of elements.</h2>

      </details>
    "
`;

exports[`collapse 2`] = `
"
        <details
          open=!{
            true ||
            [].find(link => (
              link === ((this && this.props) || props).currentPage ||
              (((this && this.props) || props).currentPage === process.env.baseURL &&
                link === window.configuration.currentFirstPage.replace('.md', '.html'))
              )
            )
          !}
        >
          <summary className='collapse-summary'>
            <p></p>
          </summary>
      <h1>Title</h1>
<h2>I can contain various combinations of elements.</h2>

      </details>
    "
`;

exports[`collapse 3`] = `
"<ul>
<li>

        <details
          open=!{
            false ||
            [].find(link => (
              link === ((this && this.props) || props).currentPage ||
              (((this && this.props) || props).currentPage === process.env.baseURL &&
                link === window.configuration.currentFirstPage.replace('.md', '.html'))
              )
            )
          !}
        >
          <summary className='collapse-summary'>
            <p><a href=\\"./pages/Publishing.md\\">:package: Publishing</a></p>
          </summary>
      <ul>
<li><a href=\\"./pages/Publishing.md#setup-branch\\">Setup</a></li>
<li><a href=\\"./pages/Publishing.md#Continuous-integration\\">Continuous Integration</a>u</li>
</ul>

      </details>
    </li>
</ul>
"
`;
